"""
Snakefile for running phage assembly from Hecatomb QC'd reads.

Rachel Rodgers, June 2021
"""

import os
import sys

#----- Snakemake Set Up -----#

configfile: "./config/phage_pipeline_config.yaml"

# Paths
HECATOMB_DIR = config["Paths"]["Hecatomb"]

#----- Collect the Input Files -----#

# Pull sample names from the Hecatomb step 7 QC'd files and store in list
SAMPLES, = glob_wildcards(os.path.join(HECATOMB_DIR, "results", "QC", "step_7", "{sample}_R1.s7.out.fastq"))

#----- Rules -----#

rule all:
	input:
		os.path.join("results", "contig_abundance_table", "contig_abundance_table.txt"),
		os.path.join("results", "mmseqs_NT_results", "NT_resultsDB.m8")
		
#----- Snakemake Workflow -----#

#----- Build Contig Dictionary -----#
include: "./rules/build_contig_dictionary.smk"

#----- Build Reference From Contig Dictionary for Quantification By Mapping -----#
include: "./rules/quantify_contig_dictionary.smk"

#----- Translated Search of Contig Dictionary against the Human Gut Virome DB (GVD) using MMSeqs2 (equivalent to tblastx) -----#
include: "./rules/query_GVD.smk"

#----- Translated Search Contig Dictionary against the Gut Phage DB (GPD) using MMSeqs2 (equivalent to tblastx) -----#
include: "./rules/query_GPD.smk"

#----- Translated Search Contig Dictionary against the Metagenomic Gut Virus Catalog (MGV) using MMSeqs2 (equivalent to tblastx)-----#
include: "./rules/query_MGV.smk"

#----- Select Probable Phage Contigs from the MMSeqs Results -----#
include: "./rules/select_phage_contigs_for_NT.smk"

#----- Extract the Probable Phage Contig Sequences to a New File to Query against NT with MMSeqs2 -----#
include: "./rules/extract_contigs_for_NT.smk"

#----- Untranslated (blastn) MMSeqs2 Search of Probable Phage Contigs against NT -----#
include: "./rules/query_NT.smk"
