"""
Snakefile for running phage assembly from Hecatomb QC'd reads.

Rachel Rodgers, June 2021
"""

import os
import sys

#----- Snakemake Set Up -----#

configfile: "./config/phage_pipeline_config.yaml"

# Paths
HECATOMB_DIR = config["Paths"]["Hecatomb"]

#----- Collect the Input Files -----#

# Pull sample names from the Hecatomb step 7 QC'd files and store in list
SAMPLES, = glob_wildcards(os.path.join(HECATOMB_DIR, "results", "QC", "step_7", "{sample}_R1.s7.out.fastq"))

#----- Rules -----#

rule all:
	input:
		os.path.join("results", "GPD_tBLASTx_results", "GPD_contig_dictionary_tBLASTx_results.tab"),
		os.path.join("results", "GVD_tBLASTx_results", "GVD_contig_dictionary_tBLASTx_results.tab"),
		os.path.join("results", "ref", "genome", "2", "summary.txt"),
		os.path.join("results", "contig_abundance_table", "contig_abundance_table.txt")
		
#----- Snakemake Workflow -----#

#----- Build Contig Dictionary -----#
include: "./rules/build_contig_dictionary.smk"

#----- Build Reference From Contig Dictionary for Quantification By Mapping -----#
include: "./rules/quantify_contig_dictionary.smk"

#----- tBLASTx Contig Dictionary against the Human Gut Virome DB (GVD) -----#
include: "./rules/query_GVD.smk"

#----- tBLASTx Contig Dictionary against the Gut Phage DB (GPD) -----#
include: "./rules/query_GPD.smk"
